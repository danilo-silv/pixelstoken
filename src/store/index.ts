export { useSheetBuyProductStore } from './useSheetBuyProduct';
export { useCartStore } from './useCart';
