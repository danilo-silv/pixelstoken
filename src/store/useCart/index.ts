import { create } from 'zustand';

import { NFTData } from '@/hooks/useGetProducts/types';

interface CartState {
  products: NFTData[];
  count: number;
  addProduct: (product: NFTData) => void;
  removeProduct: (productName: string) => void;
}

export const useCartStore = create<CartState>((set) => ({
  products: [],
  count: 0,

  addProduct: (newProduct: NFTData) =>
    set((state) => {
      const productExists = state.products.some((product) => product.name === newProduct.name);

      if (!productExists) {
        const newProducts = [...state.products, newProduct];
        return {
          products: newProducts,
          count: newProducts.length,
        };
      }
      return state;
    }),

  removeProduct: (productName: string) =>
    set((state) => {
      const newProducts = state.products.filter((product) => product.name !== productName);
      return {
        products: newProducts,
        count: newProducts.length,
      };
    }),
}));
