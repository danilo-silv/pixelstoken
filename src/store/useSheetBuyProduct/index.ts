import { create } from 'zustand';

import { NFTData } from '@/hooks/useGetProducts/types';

interface SheetBuyProductState {
  isOpen: boolean;
  product?: NFTData;
  toggleIsOpen: () => void;
  setProduct: (product?: NFTData) => void;
}

export const useSheetBuyProductStore = create<SheetBuyProductState>((set) => ({
  isOpen: false,
  product: undefined,

  toggleIsOpen: () =>
    set((state) => {
      const newState = { isOpen: !state.isOpen, product: state.product };

      if (state.isOpen) {
        newState['product'] = {} as NFTData;
      }
      return newState;
    }),

  setProduct: (product?: NFTData) => set({ product }),
}));
