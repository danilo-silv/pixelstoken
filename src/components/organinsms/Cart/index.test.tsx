import { render } from '@test-utils';
import React from 'react';

import { Cart } from '.';

import { useCartStore } from '@/store';

jest.mock('@/store', () => ({
  useCartStore: jest.fn(),
}));

jest.mock('@/components/molecules/ListCart', () => {
  const { Text } = require('react-native');
  return {
    ListCart: () => <Text testID="list-cart">ListCart</Text>,
  };
});

jest.mock('../EmptyCart', () => {
  const { Text } = require('react-native');
  return {
    EmptyCart: () => <Text testID="empty-cart">EmptyCart</Text>,
  };
});

describe('Cart Component', () => {
  const setupMocks = (products: any[], count: number) => {
    (useCartStore as unknown as jest.Mock).mockReturnValue({ products, count });
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render Shopping Cart title', () => {
      setupMocks([], 0);
      const { getByText } = render(<Cart />);
      expect(getByText('Shopping Cart')).toBeTruthy();
    });

    describe('Rendering Icon FontAwesome', () => {
      it('should render FontAwesome icon correctly', () => {
        const { getByTestId, toJSON } = render(<Cart />);
        const icon = getByTestId('shopping-cart-icon');

        expect(toJSON()).toMatchSnapshot();

        expect(icon).toBeTruthy();
      });
    });

    it('should render ShoppingCart icon with given size and color', () => {
      setupMocks([], 0);
      const { getByTestId } = render(<Cart />);

      const icon = getByTestId('shopping-cart-icon');

      expect(icon.props.width).toBe(32);
      expect(icon.props.height).toBe(32);
      expect(icon.props.stroke).toBe('white');
      expect(icon.props.fill).toBe('white');
    });

    it('should render the item count if greater than 0', () => {
      setupMocks([], 5);
      const { getByText } = render(<Cart />);
      expect(getByText('5')).toBeTruthy();
    });
  });

  describe('Conditional Rendering', () => {
    it('should render ListCart if count is greater than 0', () => {
      const products = [{ name: 'Product 1' }, { name: 'Product 2' }];
      setupMocks(products, 2);
      const { getByTestId } = render(<Cart />);
      expect(getByTestId('list-cart')).toBeTruthy();
    });

    it('should render EmptyCart if count is 0', () => {
      setupMocks([], 0);
      const { getByTestId } = render(<Cart />);
      expect(getByTestId('empty-cart')).toBeTruthy();
    });
  });
});
