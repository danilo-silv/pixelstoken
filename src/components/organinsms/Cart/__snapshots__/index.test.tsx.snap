// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component Rendering Rendering Icon FontAwesome should render FontAwesome icon correctly 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "gap": 56,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={[]}
      />
      <Text
        style={
          {
            "color": "#fff",
            "fontFamily": "Inter",
            "fontSize": 24,
            "fontWeight": 700,
          }
        }
        suppressHighlighting={true}
      >
        Shopping Cart
      </Text>
      <View
        style={
          {
            "position": "relative",
          }
        }
      >
        <RNSVGSvgView
          align="xMidYMid"
          bbHeight={32}
          bbWidth={32}
          fill="white"
          focusable={false}
          height={32}
          meetOrSlice={0}
          minX={0}
          minY={0}
          stroke="white"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "color": "white",
              },
              {
                "flex": 0,
                "height": 32,
                "width": 32,
              },
            ]
          }
          testID="shopping-cart-icon"
          tintColor="white"
          vbHeight={24}
          vbWidth={24}
          width={32}
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth="2"
          >
            <RNSVGCircle
              cx="8"
              cy="21"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                ]
              }
              r="1"
              stroke={
                {
                  "payload": 4294967295,
                  "type": 0,
                }
              }
            />
            <RNSVGCircle
              cx="19"
              cy="21"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                ]
              }
              r="1"
              stroke={
                {
                  "payload": 4294967295,
                  "type": 0,
                }
              }
            />
            <RNSVGPath
              d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
              propList={
                [
                  "stroke",
                ]
              }
              stroke={
                {
                  "payload": 4294967295,
                  "type": 0,
                }
              }
            />
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
  <Text
    testID="empty-cart"
  >
    EmptyCart
  </Text>
</View>
`;
